packages/demos/stories/TableRowsDragDrop.tsx

```diff
@@ -0,0 +1,206 @@
+/**
+ * SPDX-FileCopyrightText: © 2020 Liferay, Inc. <https://liferay.com>
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import ClayTable from '@clayui/table';
+import {XYCoord} from 'dnd-core';
+import React from 'react';
+import {DropTargetMonitor, useDrag, useDrop} from 'react-dnd';
+
+interface IDraggableTableRowProps {
+	/**
+	 * Content of the row
+	 */
+	content: {
+		country: string;
+		description: string;
+		id: string;
+		region: string;
+		title: string;
+	};
+
+	/**
+	 * Index of the List Item in order to keep track of it's positioning within the array
+	 */
+	index: number;
+
+	/**
+	 * Function that handles the moving of List Items
+	 */
+	onMove: (dragIndex: number, hoverIndex: number) => void;
+}
+
+interface IDragItem {
+	/**
+	 * List Item unique identifier
+	 */
+	id: string;
+
+	/**
+	 * Index of the List Item in order to keep track of it's positioning within the array
+	 */
+	index: number;
+
+	/**
+	 * Type of data that is allowed to be dragged
+	 */
+	type: string;
+}
+
+const DraggableTableRow: React.FunctionComponent<IDraggableTableRowProps> = ({
+	content,
+	index,
+	onMove,
+}) => {
+	const ref = React.useRef<HTMLTableRowElement>(null);
+
+	const [, drop] = useDrop({
+		accept: 'tableRow',
+		hover(item: IDragItem, monitor: DropTargetMonitor) {
+			const dragIndex = item.index;
+
+			const hoverIndex = index;
+
+			if (!ref.current || dragIndex === hoverIndex) {
+				return;
+			}
+
+			const hoverBoundingRect = ref.current!.getBoundingClientRect();
+
+			const verticalMiddle =
+				(hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;
+
+			const mousePosition = monitor.getClientOffset();
+
+			const pixelsToTop =
+				(mousePosition as XYCoord).y - hoverBoundingRect.top;
+
+			const draggingUpwards =
+				dragIndex > hoverIndex && pixelsToTop > verticalMiddle * 1.5;
+
+			const draggingDownwards =
+				dragIndex < hoverIndex && pixelsToTop < verticalMiddle / 2;
+
+			if (draggingDownwards || draggingUpwards) {
+				return;
+			}
+
+			onMove(dragIndex, hoverIndex);
+
+			item.index = hoverIndex;
```

Comment:
This seems odd to be mutating the item itself. Is this the components props we are mutating? I would expect that the `onMove` callback should be sufficient for our logic.

Reply:
That line is taken straight from their examples: [Sortable Example](https://react-dnd.github.io/react-dnd/examples/sortable/simple), if you go into the `Card.tsx` file in the typescript example, on line 72 there's the following comment: 

> Note: we're mutating the monitor item here! Generally it's better to avoid mutations, but it's good here for the sake of performance to avoid expensive index searches.

@wincent what do you think, should I invest some time and try to remove the need for the mutation or keep it since it is recommended by the author of the package?

Reply:
Unfortunately, GitHub pages is down so I can't even look at the example. 😂

<img width="348" alt="https___react-dnd_github_io_react-dnd_examples_sortable_simple" src="https://user-images.githubusercontent.com/7074/78669807-1d208f00-78dd-11ea-9293-56b0242d5c3f.png">

I don't know what the `hover` property in `useDrop` actually works. It does seem like you could do the work inside `onMove`. As this code is written it looks like `onMove` takes the item from `dragIndex` and moves it to `hoverIndex`, and then this line updates the dragged item's index, which seems wrong.

eg: if I drag "E" to the "C"

```
item    A B C D E F G
index   1 2 3 4 5 6 7
```

the desired end state should be:

```
item    A B E C D F G
index   1 2 3 4 5 6 7
```

but it will be:

```
item    A B E C D F G
index   1 2 3 3 4 6 7
```

Because we're only changing the index of item "E", and not the indices of "C" and "D", which both got bumped further to the right.

I don't know whether that leads to any actual breakage though, but it might be worth doing it without mutation in `onMove()`. The simplest way to do that is probably with two passes: one to put the items in the right place, and one to adjust the index of any item that isn't right; something like (untested):

```js
	const onMove = React.useCallback(
		(dragIndex: number, hoverIndex: number) => {
			const list = [...tableColumns];

			list.splice(dragIndex, 1);

			list.splice(hoverIndex, 0, tableColumns[dragIndex]);

			setTableColumns(list.map((item, i) => {
				if (item === i) {
					return item;
				} else {
					return {
						...item,
						index: i,
					};
				}
			}));
		},
		[tableColumns]
	);
```

Reply:
You could do this in one pass, but I think it would be fiddly to get right, and harder to read. We don't expect the number of columns to be very large (even in crazy scenarios, I can't see it being more than about a dozen or so), so I would be wary of prematurely optimizing.